sw/airborne/modules/loggers/sd_logger_spi_direct.c

```diff
@@ -0,0 +1,388 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with paparazzi; see the file COPYING.  If not, write to
+ * the Free Software Foundation, 59 Temple Place - Suite 330,
+ * Boston, MA 02111-1307, USA.
+ *
+ */
+
+/** @file modules/loggers/sd_logger.c
+ *  @brief Module for real time logging using an SD Card in SPI mode.
+ * This module buffers raw data from IMU and control inputs into a buffer of the SDCard peripheral.
+ * Once the buffer is full, the entire data block is written to the card. No filesystem like FAT is used.
+ * The card can be soldered directly to the SPI bus, so the increase in mass is minimal.
+ * The logging is started and stopped by an external application in the ground station.
+ * After logging, the data can be downloaded over telemetry messages using the same application in the ground station.
+ *
+ * The FIXME's in this file are discovered after a successful test of the module. They are not critical,
+ * but will be fixed and tested in a next release.
+ */
+
+#include "peripherals/sdcard_spi.h"
+#include "subsystems/datalink/telemetry.h"
+
+#ifdef TEST
+/* The original messages.h uses inline functions, which are incompatible with cmock. TEST is defined by the unittest framework */
+#include "messages_testable.h"
+#endif // TEST
+
+#include "subsystems/imu.h"
+#include "subsystems/actuators/actuators_pwm_arch.h"
+#include "sd_logger_spi_direct.h"
+
+/* Definition of the sdlogger */
+struct SdLogger sdlogger;
+
+/**
+ * @brief Start function called during initialization of the autopilot.
+ * It starts the initialization procedure of the SDCard.
+ */
+void sd_logger_start(void)
+{
+  sdcard_spi_init(&sdcard1, &(SD_LOGGER_SPI_LINK_DEVICE), SD_LOGGER_SPI_LINK_SLAVE_NUMBER);
+  sdlogger.status = SdLogger_Initializing;
+}
+
+/**
+ * @brief Periodic function of the logger, called at modules main frequency.
+ */
+void sd_logger_periodic(void)
+{
+  /* Keep the SDCard running at the same rate */
+  sdcard_spi_periodic(&sdcard1);
+
+  switch (sdlogger.status) {
+
+      /* SD card is initializing, check to see if it is ready */
+    case SdLogger_Initializing:
+      if (sdcard1.status == SDCard_Error) {
+        sdlogger.status = SdLogger_Error;
+      } else if (sdcard1.status == SDCard_Idle) {
+        sdlogger.status = SdLogger_Idle;
+      }
+      /* else: state unchanged */
+      break;
+
+      /* Logging data, write data to buffer */
+    case SdLogger_Logging:
+      sdlogger.packet_count++;
+      sd_logger_uint32_to_buffer(sdlogger.packet_count, &sdcard1.output_buf[SD_LOGGER_BUFFER_OFFSET + sdlogger.buffer_addr]);
+      sd_logger_int32_to_buffer(imu.accel_unscaled.x,
+                                &sdcard1.output_buf[SD_LOGGER_BUFFER_OFFSET + sdlogger.buffer_addr + 4]);
```

Comment:
All this part is very user specific. Can't we have something more generic with a buffer filled by an external user module ?